src/v/raft/recovery_stm.cc

```diff
@@ -218,50 +218,58 @@ recovery_stm::read_range_for_recovery(
-        _stop_requested = true;
-        co_return std::nullopt;
-    }
-    vlog(
-      _ctxlog.trace,
-      "Read batches in range [{},{}] for recovery",
-      batches.front().base_offset(),
-      batches.back().last_offset());
-
-    auto gap_filled_batches = details::make_ghost_batches_in_gaps(
-      start_offset, std::move(batches));
-    _base_batch_offset = gap_filled_batches.begin()->base_offset();
-    _last_batch_offset = gap_filled_batches.back().last_offset();
-
-    if (is_learner && _ptr->_recovery_throttle) {
-        const auto size = std::accumulate(
-          gap_filled_batches.cbegin(),
-          gap_filled_batches.cend(),
-          size_t{0},
-          [](size_t acc, const auto& batch) {
-              return acc + batch.size_bytes();
-          });
+        if (batches.empty()) {
+            vlog(_ctxlog.trace, "Read no batches for recovery, stopping");
+            _stop_requested = true;
+            co_return std::nullopt;
+        }
         vlog(
           _ctxlog.trace,
-          "Requesting throttle for {} bytes, available in throttle: {}",
-          size,
-          _ptr->_recovery_throttle->get().available());
-        co_await _ptr->_recovery_throttle->get()
-          .throttle(size, _ptr->_as)
-          .handle_exception_type([this](const ss::broken_semaphore&) {
-              vlog(_ctxlog.info, "Recovery throttling has stopped");
-          });
-    }
+          "Read batches in range [{},{}] for recovery",
+          batches.front().base_offset(),
+          batches.back().last_offset());
+
+        auto gap_filled_batches = details::make_ghost_batches_in_gaps(
+          start_offset, std::move(batches));
+        _base_batch_offset = gap_filled_batches.begin()->base_offset();
+        _last_batch_offset = gap_filled_batches.back().last_offset();
+
+        if (is_learner && _ptr->_recovery_throttle) {
+            const auto size = std::accumulate(
+              gap_filled_batches.cbegin(),
+              gap_filled_batches.cend(),
+              size_t{0},
+              [](size_t acc, const auto& batch) {
+                  return acc + batch.size_bytes();
+              });
+            vlog(
+              _ctxlog.trace,
+              "Requesting throttle for {} bytes, available in throttle: {}",
+              size,
+              _ptr->_recovery_throttle->get().available());
+            co_await _ptr->_recovery_throttle->get()
+              .throttle(size, _ptr->_as)
+              .handle_exception_type([this](const ss::broken_semaphore&) {
+                  vlog(_ctxlog.info, "Recovery throttling has stopped");
+              });
+        }
 
-    co_return model::make_foreign_memory_record_batch_reader(
-      std::move(gap_filled_batches));
+        co_return model::make_foreign_memory_record_batch_reader(
+          std::move(gap_filled_batches));
+    } catch (const ss::timed_out_error& e) {
+        vlog(
+          _ctxlog.error,
```

Comment:
Recovery will be retried after this error, right? If so, what is the rationale for logging it at error severity? So that we can catch it in tests?

Reply:
exactly, i would love this to be as visible as possible.